internal/router/router.go

```diff
@@ -151,17 +167,23 @@ type Router struct {
 // The exists function may be called to test whether an entity
 // exists when an API endpoint needs to know that
 // but has no appropriate handler to call.
+//
+// The authorizeAdmin function will be called to authorize admin
+// requests. The Cause of the authorizeAdmin error will be left
+// unchanged, as for the handlers.
 func New(
 	handlers *Handlers,
 	resolveURL func(id *charm.Reference) error,
 	authorize func(id *charm.Reference, req *http.Request) error,
 	exists func(id *charm.Reference, req *http.Request) (bool, error),
+	authorizeWithPerms func(req *http.Request, read, write []string) error,
```

Comment:
permsForEntity should probably return a map, no? But I agree that what you suggest leads to a more flexible approach.